ASoC: sophgo: add CV1800 AIAO mux and I2S support

stavinsky opened the pull request:
This patch series adds basic audio support for Sophgo CV1800B, as used on boards such as the Milk-V Duo. The series introduces an AIAO audio mux driver, an I2S controller driver, corresponding DT bindings, and DTS updates to wire the components together.

The implementation is based on vendor documentation and testing on real hardware. This series relies on recent fixes in the DesignWare AXI DMA support; in particular, correct operation depends on the DMA changes discussed at:
https://lore.kernel.org/all/20251214224601.598358-1-inochiama@gmail.com/ The current driver implementation supports a fixed audio configuration of 48 kHz sample rate. The series has been tested on the Milk-V Duo and Milk-V Duo 256M boards using the Sophgo SG2002 SoC. The implementation is expected to also work on Milk-V Duo boards based on the SG2000 SoC, as the audio and DMA blocks are closely related. 

Known hardware limitation:
On CV1800B / SG2002, the I2S2 output pins cannot be enabled via pinctrl alone. Due to SoC design constraints, the output path becomes functional only after additional vendor-specific register programming. This series makes the limitation explicit and does not attempt to work around it implicitly via pinctrl or undocumented behavior.

inochisa commented on drivers/mux/Kconfig:
```diff
@@ -59,4 +59,12 @@ config MUX_MMIO
+config MUX_CV1800B_AIAO
+    tristate "Sophgo CV1800B AIAO mux support"
+    depends on SND_SOC && OF
+	depends on MULTIPLEXER
```
Possible unaligned depends on?

stavinsky replied:
Don't get it. You mean style of the MULTIPLEXER is not needed? I'm using consumer.h from this module. It will not compile without it.

inochisa replied:
These two "DEPENDS ON" line should be aligned (at least it is not aligned in the github web), or just put them in one line.

stavinsky replied:
identation, fixed

inochisa commented on drivers/mux/cv1800b-aiao-mux.c:
```diff
@@ -0,0 +1,229 @@
+	.reg_stride = 4,
+	.val_bits = 32,
+};
+/* AIAO fields */
```
One blank line before the comment, also for the below

stavinsky replied:
done.

inochisa commented on drivers/mux/cv1800b-aiao-mux.c:
```diff
@@ -0,0 +1,229 @@
+#define CV1800B_AUDIO_PDM_CTRL 0x040
+#define CV1800B_I2S_SYS_INT_EN 0x060
+#define CV1800B_I2S_SYS_INTS 0x064
+enum cv1800b_aiao_fields {
```
One blank line before

inochisa commented on drivers/mux/cv1800b-aiao-mux.c:
```diff
@@ -0,0 +1,229 @@
+	/* End of register map */
+	CV1800B_AIAO_MAX_FIELDS,
+};
+static const struct reg_field cv1800b_aiao_reg_fields[] = {
```
one blank line before

inochisa commented on drivers/mux/cv1800b-aiao-mux.c:
```diff
@@ -0,0 +1,229 @@
+#define CV1800B_I2S_BCLK_OEN_SEL 0x030
+#define CV1800B_AUDIO_PDM_CTRL 0x040
+#define CV1800B_I2S_SYS_INT_EN 0x060
+#define CV1800B_I2S_SYS_INTS 0x064
```
Align the register offset number is better

stavinsky replied:
done

inochisa commented on drivers/mux/cv1800b-aiao-mux.c:
```diff
@@ -0,0 +1,229 @@
+	SIG_FS_IN, /* 4..7  */
+	SIG_SDI_IN, /* 8..11 */
+	SIG_SDO_OUT, /* 12..15 */
+	SIG_MAX,
```
Align the comment here

stavinsky replied:
done

inochisa commented on drivers/mux/cv1800b-aiao-mux.c:
```diff
@@ -0,0 +1,229 @@
+	[SIG_FS_IN] = CV1800B_I2S_TDM_FS_IN_SEL,
+	[SIG_SCLK_IN] = CV1800B_I2S_TDM_SCLK_IN_SEL,
+	[SIG_SDI_IN] = CV1800B_I2S_TDM_SDI_IN_SEL,
+	[SIG_SDO_OUT] = CV1800B_I2S_TDM_SDO_OUT_SEL,
```
Align the assigne

stavinsky replied:
done

inochisa commented on drivers/mux/cv1800b-aiao-mux.c:
```diff
@@ -0,0 +1,229 @@
+	struct regmap *map;
+	struct regmap_field *aiao_fields[CV1800B_AIAO_MAX_FIELDS];
+};
+static int cv1800b_mux_set(struct mux_control *mux, int state)
```
one blank line before

inochisa commented on drivers/mux/cv1800b-aiao-mux.c:
```diff
@@ -0,0 +1,229 @@
+};
+static void cv1800b_aiao_mux_remove(struct platform_device *pdev)
+{
+}
```
Drop this.

stavinsky replied:
done

inochisa commented on drivers/mux/cv1800b-aiao-mux.c:
```diff
@@ -0,0 +1,229 @@
+module_platform_driver(cv1800b_aiao_mux_driver);
+
+MODULE_DESCRIPTION("Sophgo cv1800b I2S/TDM Multiplexor driver");
+MODULE_AUTHOR("Anton D. Stavinskii");
```
Add you email here.

stavinsky replied:
done

inochisa commented on Documentation/devicetree/bindings/sound/cvitek,cv1800b-i2s.yaml:
```diff
@@ -0,0 +1,188 @@
+# SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+%YAML 1.2
+---
+$id: http://devicetree.org/schemas/sound/cvitek,cv1800b-i2s.yaml#
```
The filename should be `sophgo,cv1800b-i2s.yaml`, no cvitek anymore

stavinsky replied:
done

inochisa commented on arch/riscv/boot/dts/sophgo/cv180x.dtsi:
```diff
@@ -5,6 +5,7 @@
  */
 
 #include <dt-bindings/clock/sophgo,cv1800.h>
+#include <dt-bindings/sound/cv1800b-aiao.h>
```
Move this include after `#include <dt-bindings/interrupt-controller/irq.h>`

stavinsky replied:
done

inochisa commented on arch/riscv/boot/dts/sophgo/cv180x.dtsi:
```diff
@@ -447,7 +448,75 @@
+			reg = <0x04108000 0x1000>;
+			#mux-state-cells = <2>;
+		};
+		i2s0: i2s@4100000 {
```
one blank line before you node. Also for another nodes.

stavinsky replied:
done

inochisa commented on arch/riscv/boot/dts/sophgo/sg2002-milkv-duo-256m.dts:
```diff
@@ -0,0 +1,245 @@
+			power-source = <3300>;
+		};
+	};
+};
```
merge this with the pinctrl node in line35

stavinsky replied:
the file will be removed. It is just for testing.

stavinsky replied:
removed completely to not confuse anyone

inochisa commented on arch/riscv/boot/dts/sophgo/sg2002-milkv-duo-256m.dts:
```diff
@@ -0,0 +1,245 @@
+	pinctrl-names = "default";
+	status = "okay";
+};
+/// spi  begin
```
No C++ style comments, If you do want one, just use `/* */`

stavinsky replied:
the file will be removed. It is just for internal testing.

inochisa commented on arch/riscv/boot/dts/sophgo/sg2002-milkv-duo-256m.dts:
```diff
@@ -0,0 +1,245 @@
+		compatible = "sophgo,sg2002-ephy";
+	};
+
+};
```
What's this?

stavinsky replied:
https://github.com/sophgo/sophgo-doc/issues/174#event-21395297524
it is this magic. 
here the dummy driver I had to write to be able to input or output to i2s2 pins. 
https://github.com/stavinsky/milk_duo_256-buildroot/blob/i2s_capture/br-external/package/ephy-i2s2/src/ephy.c
I'm sorry I mention this not very obvious but this DT file is just for you to have an idea how ti works. I will delete it from the patch because it enables staff that probably not needed for the end user.

stavinsky replied:
I'm really interested in you thoughts here.

inochisa replied:
This is a internal thing for ethernet phy. You do not need it. and we prefer controll it through MDIO.

inochisa commented on Documentation/devicetree/bindings/sound/sophgo,cv1800b-i2s.yaml:
```diff
@@ -0,0 +1,188 @@
+    maxItems: 1
+
+  assigned-clock-rates:
+    maxItems: 1
```
IIRC, this is not necessary

stavinsky replied:
that is the question. Yes it is not required and dts_check will pass without it but how can I tell to the reader that this has to be defined for proper work?

inochisa replied:
These two properties are defined globally, which means all the node has this two, you can just use it.

stavinsky replied:
removed completely. It is not required anymore

inochisa commented on Documentation/devicetree/bindings/sound/cvitek,cv1800b-i2s.yaml:
```diff
@@ -0,0 +1,188 @@
+  - Anton D. Stavinskii <stavinsky@gmail.com>
+
+description: |
+  I2S/TDM controller found in Cvitek CV1800B / Sophgo SG2002/SG2000 SoCs.
```
No Cvitek

stavinsky replied:
done. now Sophgo

inochisa commented on Documentation/devicetree/bindings/sound/cvitek,cv1800b-i2s.yaml:
```diff
@@ -0,0 +1,188 @@
+  dma-names:
+    items:
+      - enum: [ rx, tx ]
+      - enum: [ rx, tx ]
```
IIRC, both rx and tx are required, so just use items for it

stavinsky replied:
we have only 8 dmas so we need to add to the user ability to not consume dma without need. TDM can be only rx, tx or both or nothing (for example if it is clock source only )
fixed this to allow no rx tx defined at all

inochisa replied:
OK, lets leave it.

inochisa commented on Documentation/devicetree/bindings/mux/sophgo,cv1800b-aiao-mux.yaml:
```diff
@@ -0,0 +1,40 @@
+# SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+%YAML 1.2
+---
+$id: http://devicetree.org/schemas/mux/sophgo,cv1800b-aiao-mux.yaml#
```
How about moving this into sound and refactor to use dai-common.yaml?
You can reference `Documentation/devicetree/bindings/sound/simple-audio-mux.yaml`

stavinsky replied:
very good Idea, will do but it will take some time. I will move all the parameters i can to be more generic and compatible with simple-audio-card.

stavinsky replied:
I've checked the possibility to use simple-audio-mux.yaml and I could find a way how to use it. Looks like it suppose to control gpio but in our case we are changing interconnections between tdm to tdm, and tdm to IO
But I'm trying to make tdm cpus more compatible with simple-audio-card Documentation/devicetree/bindings/sound/simple-card.yaml. it looks like good idea and doable.

inochisa commented on Documentation/devicetree/bindings/sound/cvitek,cv1800b-i2s.yaml:
```diff
@@ -0,0 +1,188 @@
+
+  aiao:
+    $ref: /schemas/types.yaml#/definitions/phandle
+    description: Phandle to AIAO block used for internal audio routing/control.
```
I think this is the multiplexer, right? If so, using something like `sophgo,multiplexer`?

stavinsky replied:
i forgot to remove it. It is not needed anymore with the current revision

stavinsky replied:
done

inochisa reviewed (changes requested):
I think this is a first step to review. I suggest you should check you patch format by executing `scripts/checkpatch.pl` with `--strict` parameter. I will continue to review your logic after this change. Thanks.

stavinsky commented:
> I think this is a first step to review. I suggest you should check you patch format by executing `scripts/checkpatch.pl` with `--strict` parameter. I will continue to review your logic after this change. Thanks.
`check_dtbs` also passes

Thank you for pointing me to `checkpatch.pl`. Fixed everything i could so far. 
changes in new update: 
assigned clock refactored - now it is calculated value inside driver
removed master flag from dts, now it is respecting simple-audio-card parameters you can see example in provided dts
I aslo don't like my mux approach but still can't find better solution.

Sorry for speding you time on style problems. Will try to avoid such things in future. 
Thank you for you time and review.  Really appreciate all your help.

inochisa commented:
Good update, and I think you should do the additional check for your patches:
1. `make DT_CHECKER_FLAGS=-m dt_binding_check`
2. `make DT_CHECKER_FLAGS=-m W=1 dtbs_check`

inochisa commented on Documentation/devicetree/bindings/sound/sophgo,cv1800b-i2s.yaml:
```diff
@@ -0,0 +1,158 @@
+        status = "okay";
+    };
+
+...
```
Leave an empty line at the end. This will make git confused.

inochisa commented on Documentation/devicetree/bindings/sound/sophgo,cv1800b-i2s.yaml:
```diff
@@ -0,0 +1,158 @@
+        mux-states = <&aiao CV1800B_AIAO_FS(3) 2>,
+        mux-state-names = "fs-in", "sclk-in", "sdi-in";
+        status = "okay";
+    };
```
Check: https://docs.kernel.org/devicetree/bindings/dts-coding-style.html
And I do not think all are needed, The last one is enough.

stavinsky replied:
The reason of second example was because it describes the working way to achieve rx/tx simultaneous capability on the same physical port. tdm2 is clock only. tdm1/tdm3 are rx/tx. 
But I agree it looks redundant. In most of the case user will take first example. I will remove.

stavinsky replied:
removed

inochisa commented on sound/soc/Kconfig:
```diff
@@ -139,6 +139,7 @@ source "sound/soc/uniphier/Kconfig"
 source "sound/soc/ux500/Kconfig"
 source "sound/soc/xilinx/Kconfig"
 source "sound/soc/xtensa/Kconfig"
+source "sound/soc/sophgo/Kconfig"
```
Follow alphabetical order

stavinsky replied:
done

inochisa commented on sound/soc/Makefile:
```diff
@@ -83,3 +83,4 @@ obj-$(CONFIG_SND_SOC)	+= ux500/
 obj-$(CONFIG_SND_SOC)	+= xilinx/
 obj-$(CONFIG_SND_SOC)	+= xtensa/
 obj-$(CONFIG_SND_SOC)	+= sdw_utils/
+obj-$(CONFIG_SND_SOC)   += sophgo/
```
Follow alphabetical order

stavinsky replied:
done

inochisa commented on sound/soc/sophgo/Kconfig:
```diff
@@ -0,0 +1,9 @@
+config SND_SOC_CV1800B_TDM
+    tristate "Sophgo CV1800B I2S/TDM support"
+    depends on SND_SOC && OF
+    select SND_SOC_GENERIC_DMAENGINE_PCM
+    select MULTIPLEXER
+    help
+      Say Y or M here to enable I2S/TDM support on Sophgo CV1800B SoCs.
+
+
```
It is so hard for me to give a comment, so I give you an example for reference. Also, suggest for detailed help if possible.
```
# SPDX-License-Identifier: GPL-2.0-only

menu "Sophgo"
	depends on COMPILE_TEST || ARCH_SOPHGO

config SND_SOC_CV1800B_TDM
    tristate "Sophgo CV1800B I2S/TDM support"
    depends on SND_SOC && OF
    select SND_SOC_GENERIC_DMAENGINE_PCM
    select MULTIPLEXER
    help
      Say Y or M here to enable I2S/TDM support on Sophgo CV1800B SoCs.

endmenu
```

stavinsky replied:
done

inochisa commented on sound/soc/sophgo/Makefile:
```diff
@@ -0,0 +1,2 @@
+
+obj-$(CONFIG_SND_SOC_CV1800B_TDM) += cv1800b_tdm.o
```
```
# SPDX-License-Identifier: GPL-2.0
# ROCKCHIP Sophgo Support

obj-$(CONFIG_SND_SOC_CV1800B_TDM) += cv1800b_tdm.o
```

inochisa commented on sound/soc/sophgo/cv1800b_tdm.c:
```diff
@@ -0,0 +1,619 @@
+static const struct of_device_id cv1800b_i2s_of_match[] = {
+	{
+		.compatible = "sophgo,cv1800b-i2s",
+	},
```
```
{ .compatible = "sophgo,cv1800b-i2s", }
```

stavinsky replied:
done

inochisa commented on sound/soc/sophgo/cv1800b_tdm.c:
```diff
@@ -0,0 +1,619 @@
+		return -ENODEV;
+	regs = devm_ioremap_resource(dev, res);
+	if (IS_ERR(regs))
+		return PTR_ERR(regs);
```
Use devm_platform_ioremap_resource

stavinsky replied:
done, but I still need res variable. So I had to keep regs = devm_ioremap_resource(dev, res); because for dma I need to provide register offset.

inochisa commented on sound/soc/sophgo/cv1800b_tdm.c:
```diff
@@ -0,0 +1,619 @@
+
+	i2s->regs = devm_regmap_init_mmio(dev, regs, &cv1800b_regmap_config);
+	if (IS_ERR(i2s->regs))
+		return dev_err_probe(dev, PTR_ERR(i2s->regs), "regmap init\n");
```
Is there any reason for a regmap? If no, use readl/writel instead. It is not a good thing to use regmap always.

stavinsky replied:
I was inspiring by https://docs.kernel.org/sound/soc/codec.html#codec-control-io (i know it is not directly applicable in our case) and other sound soc drivers. So I just didn't find any reasons to not use regmap if it is available and more simple to describe.
I'm new to kernel development, can you please explain why I should avoid using regmaps in future?

inochisa replied:
There are a few reasons for not using regmap:
1. It is faster and simple to use `readx` and `writex` than regmap. And regmap only supports 32 bit I/O which `readx`/`writex` support 8/16/32/64 I/O (by replacing `x` with `b|w|l|q`).
2. Avoid unnecessary check, the regmap always return a status code needed to be checked. If you use `read`/`write`, there is no need to check them.
3. Since regmap may not perform write operation if you write number is the same as it cached, which may break some devices. To avoid this, you should be more careful for it.

I think the only two cases to use a regmap are:
1. The I/O area is a syscon device.
2. The I/O always require acquiring a lock (For example, the clock driver).

stavinsky replied:
thank you for explanations. 
Please check last commits. I made it as a separate commit because I'm not sure if this is what you requesting. 
got rid of regmap, rewrote everything to iomem. 
tested everything, seems to be working.

inochisa commented on sound/soc/sophgo/cv1800b_tdm.c:
```diff
@@ -0,0 +1,619 @@
+			i2s->cap = CV1800B_I2S_CAP_TX;
+
+		else if (!strcmp(cap, "txrx"))
+			i2s->cap = CV1800B_I2S_CAP_TXRX;
```
Can we use multiple compatible to replace this ugly capability check? For example 
```
sophgo,cv1800b-i2s-tx
sophgo,cv1800b-i2s-rx
sophgo,cv1800b-i2s-rxtx
```

Although these are not very good. it is clear for user to know which one.

stavinsky replied:
may be the word capability is wrong. Each TDM can do 3 things. 
TDM can do receive, transmit, both. There is also a special case where it doesn't produce any data but used as a source for generating BCLK, LRCLK. 
So it is not description what is it but more like a request: "I want you to be a receiver" 

While answering here I figured out that I can register both directions in all the cases. Because the codec connected via simple-audio-card will be preventing to use wrong direction. So this code looks redundant. Will remove it and make the card expose both directions.

inochisa commented on sound/soc/sophgo/cv1800b_tdm.c:
```diff
@@ -0,0 +1,619 @@
+			 "tdm-id %d is set to only clock generation mode\n",
+			 i2s->tdm_id);
+		return 0;
+	}
```
I think it maybe better for using a clock driver.

stavinsky replied:
This is a special case described before. It is only needed when you want to not expose any rx/tx channels but only start lrclk/bclk. I could just remove this for simplification. Up to you. Clock driver is not enough here because I need to change this register. 
<img width="874" height="337" alt="image" src="https://github.com/user-attachments/assets/792a5861-49cb-4f37-8acd-c173b3181137" />

stavinsky replied:
removed for simplicity.

stavinsky commented:
another batch of changes. 
hope no more style problems. 
removed capability and tdm-id properties 
fixed Kconfigs. (new line, text etc)
fixed `dma-names` and in yaml. `dtbs_check` and `dt_binding_check` don't complain about my files anymore. 
i'm running `checkpatch` with --strict now.
btw, I don't know how it is better for you if I click resolve on resolved threads or you prefer to track changes and close threads by yourself.

inochisa commented:
>  I don't know how it is better for you if I click resolve on resolved threads or you prefer to track changes and close threads by yourself.

For simple thing like code refactor, you can just close the threads, for some complex thing, just leave it. And I will decide whether it is OK to close

inochisa commented:
@stavinsky Also, if you can, just do a squash merge for the last commit, it is too hard to see the change.

inochisa commented on sound/soc/sophgo/Makefile:
```diff
@@ -0,0 +1,4 @@
+
+# SPDX-License-Identifier: GPL-2.0
+# ROCKCHIP Sophgo Support
+obj-$(CONFIG_SND_SOC_CV1800B_TDM) += cv1800b_tdm.o
```
No new line at fire fist, just move it before the first obj
Also `ROCKCHIP Sophgo Support` should be `Sophgo Platform Support`

stavinsky replied:
done

stavinsky commented:
> @stavinsky Also, if you can, just do a squash merge for the last commit, it is too hard to see the change.

@inochisa Sorry i was thinking it is easier to check separate commit. Fixed

inochisa commented on sound/soc/sophgo/cv1800b_tdm.c:
```diff
@@ -0,0 +1,527 @@
+
+	tmp = u32_replace_bits(tmp, value, mask);
+	writel(tmp, base + reg);
+}
```
I think it is better to direct use `readl`/`writel`
It possible to just use one write for multiple replace.

stavinsky replied:
little bit more code, but done.

inochisa commented on sound/soc/sophgo/cv1800b_tdm.c:
```diff
@@ -0,0 +1,527 @@
+		return ret;
+	ret = cv1800b_try_select_state_optional(dev, "sdi-in");
+	if (ret)
+		return ret;
```
It is better to use a for loop.

stavinsky replied:
true, done

inochisa commented on sound/soc/sophgo/cv1800b_tdm.c:
```diff
@@ -0,0 +1,527 @@
+static const struct of_device_id cv1800b_i2s_of_match[] = {
+	{
+		.compatible = "sophgo,cv1800b-i2s",
+	},
```
Why you change this again?

stavinsky replied:
I'm fighting with formatter. It made it again because of comma after string constant. not it is is fixed. Should never be happen again. Sorry.

stavinsky commented:
all the fixes from the last review are done. 
I'm still learning alsa. Looking for way to get rid of mux driver. I agree it doesn't look good.  Now I'm reading about DAPM.  
closed all style issues threads. 
sorry for two pushes instead of  one (i'm still getting use to current workflow and I have to test everything on hardware with you last dma changes, so I accidentally pushed you those changes. )
@inochisa

inochisa commented on drivers/mux/cv1800b-aiao-mux.c:
```diff
@@ -0,0 +1,165 @@
+		break;
+	default:
+		return -EINVAL;
+	}
```
How about using `GENMASK(i * 4 + 2, i * 4)` for `TDMx_SEL_MASK`

stavinsky replied:
I would happy to find a better way too. 
your solution gives me 
./include/linux/bitfield.h:183:17: error: call to ‘__bad_mask’ declared with attribute error: bad bitfield mask
  183 |                 __bad_mask();

runtime mask is not allowed.

i could do manual if applicable
```
unsigned int shift = tdm * 4;
tmp &= ~(0x7u << shift); 
tmp |= ((u32)state & 0x7u) << shift; 

```
What do you think, @inochisa ?

inochisa replied:
Just let the `state` and `tmp` be `u32`, and use `0x7` is enough, no necessary for `0x7u`.

Otherwise LGTM.

stavinsky replied:
done

inochisa commented on drivers/mux/cv1800b-aiao-mux.c:
```diff
@@ -0,0 +1,165 @@
+	writel(tmp, reg);
+
+	dev_info(priv->dev, "mux is set for tdm: %i, value: %i, sig: %i", tdm,
+		 state, sig);
```
`dev_dbg`, this also applies to other `dev_info`.

stavinsky replied:
done dev_info -> dev_dbg in both files

inochisa commented on include/dt-bindings/sound/cv1800b-aiao.h:
```diff
@@ -0,0 +1,32 @@
+#define CV1800B_AIAO_SDI(tdm) CV1800B_AIAO_MUX(CV1800B_SIG_SDI_IN, (tdm))
+#define CV1800B_AIAO_SDO(tdm) CV1800B_AIAO_MUX(CV1800B_SIG_SDO_OUT, (tdm))
+
+#endif /* _DT_BINDINGS_SOUND_CV1800B_AIAO_H */
```
binding header should be along with binding docs.

stavinsky replied:
done

inochisa commented on arch/riscv/boot/dts/sophgo/cv180x.dtsi:
```diff
@@ -448,6 +449,60 @@
+			#size-cells = <0>;
+			clocks = <&clk CLK_APB_I2S0>,  <&clk CLK_SDMA_AUD0>;
+			clock-names = "i2s", "mclk";
+			dmas = <&dmamux 0 1>, <&dmamux 1 1>;
```
Use this file [cv180x-dmamux.h.c](https://github.com/user-attachments/files/24301085/cv180x-dmamux.h.c) to replace magic numbers. (Place it at `arch/riscv/boot/dts/sophgo/cv180x-dmamux.h`, as github does not support upload header file)

stavinsky replied:
done, `arch/riscv/boot/dts/sophgo/cv180x-dmamux.h` was put to i2s driver bindings commit for now

stavinsky replied:
moved to nodes commit

inochisa commented on arch/riscv/boot/dts/sophgo/Makefile:
```diff
@@ -6,3 +6,4 @@ dtb-$(CONFIG_ARCH_SOPHGO) += sg2042-milkv-pioneer.dtb
 dtb-$(CONFIG_ARCH_SOPHGO) += sg2042-evb-v1.dtb
 dtb-$(CONFIG_ARCH_SOPHGO) += sg2042-evb-v2.dtb
 dtb-$(CONFIG_ARCH_SOPHGO) += sg2044-sophgo-srd3-10.dtb
+dtb-$(CONFIG_ARCH_SOPHGO) += sg2002-milkv-duo-256m.dtb
```
For this, adding a separate commit for new sg2002 dtb. So I can force I2S change.

stavinsky replied:
removed from this series for now if I got you correctly along with dts

inochisa committed:
Merge branch 'dt/riscv' into for-next

inochisa commented on drivers/mux/cv1800b-aiao-mux.c:
```diff
@@ -0,0 +1,165 @@
+
+	for (int i = 0; i < mux_chip->controllers; i++)
+		mux_chip->mux[i].states = 8;
+	ret = devm_mux_chip_register(dev, mux_chip);
```
one new empty line before this.

inochisa replied:
one empty line before this line.

inochisa replied:
one empty line before this line.

inochisa replied:
one empty line before this line.

inochisa commented on drivers/mux/cv1800b-aiao-mux.c:
```diff
@@ -0,0 +1,165 @@
+	{ .compatible = "sophgo,cv1800b-aiao-mux" },
+	{}
+};
+
```
This empty should not be added.

inochisa replied:
no empty line here.

inochisa replied:
no empty line here.

inochisa commented on arch/riscv/boot/dts/sophgo/cv180x-dmamux.h:
```diff
@@ -0,0 +1,57 @@
+/* SPDX-License-Identifier: (GPL-2.0 OR MIT) */
+/*
+ * Copyright (C) 2025 Inochi Amaoto <inochiama@gmail.com>
+ */
+
```
This file is add along with the dts change, it is not a binding.

inochisa commented on sound/soc/sophgo/cv1800b_tdm.c:
```diff
@@ -0,0 +1,549 @@
+
+	int ret = clk_set_rate(i2s->clk_mclk, rate * channels * physical_width *
+						      CV1800B_BCLK_DIV *
+						      CV1800B_MCLK_DIV);
```
u32_replace_bits

stavinsky replied:
sorry, didn't get it

inochisa replied:
Oh, sorry, I have missed it. move all definition to the start of the function. It is a violation of the kernel code style.

inochisa commented on drivers/mux/cv1800b-aiao-mux.c:
```diff
@@ -0,0 +1,165 @@
+	{ .compatible = "sophgo,cv1800b-aiao-mux" },
+	{}
+};
+
```
no empty line here, this apply to all your patches

stavinsky replied:
done

inochisa commented on arch/riscv/boot/dts/sophgo/cv180x-dmamux.h:
```diff
@@ -0,0 +1,57 @@
+/* SPDX-License-Identifier: (GPL-2.0 OR MIT) */
```
This header file should go alone with the dts change. It is not a binding

stavinsky replied:
duplicate. closing

inochisa commented on arch/riscv/boot/dts/sophgo/cv180x-dmamux.h:
```diff
@@ -0,0 +1,57 @@
+/* SPDX-License-Identifier: (GPL-2.0 OR MIT) */
```
This header file should go alone with the dts change. It is not a binding

stavinsky replied:
dup

inochisa commented on sound/soc/sophgo/cv1800b_tdm.c:
```diff
@@ -0,0 +1,549 @@
+		if (ret == -EPROBE_DEFER)
+			return ret;
+		dev_err(dev, "mux state '%s' select failed: %d\n", name, ret);
+		return ret;
```
Just do the return, all are errors.

stavinsky replied:
done

inochisa commented on sound/soc/sophgo/cv1800b_tdm.c:
```diff
@@ -0,0 +1,549 @@
+
+	int ret = clk_set_rate(i2s->clk_mclk, rate * channels * physical_width *
+						      CV1800B_BCLK_DIV *
+						      CV1800B_MCLK_DIV);
```
Define everything at the start of the function, this is a requirement.

stavinsky replied:
done, if I got correctly

inochisa commented on sound/soc/sophgo/cv1800b_tdm.c:
```diff
@@ -0,0 +1,549 @@
+	u32 val;
+
+	if (!i2s->base)
+		return;
```
I do not think there is a case for that, it is just wrong.

stavinsky replied:
done

inochisa commented on sound/soc/sophgo/cv1800b_tdm.c:
```diff
@@ -0,0 +1,549 @@
+		if (ret == -EPROBE_DEFER)
+			return ret;
+		if (ret == -ENOENT || ret == -ENODEV || ret == -ENODATA)
+			return 0;
```
I do not think it is good to request things that are not required.
This is more like a abuse.

stavinsky replied:
I don't know how to avoid this. 
I either have to force user to define all mux states or do something like this. 
Driver can start without defining all mux states now.  
Should I force the user to define all mux states in bindings, @inochisa  ?
Thanks

inochisa replied:
Yeah, that is a hard things. I agree with not all mux states are required. I think we should indicate the operation mode of I2S in an acceptable way.

stavinsky replied:
which way do you prefer? I will agree with any decision.

inochisa replied:
Give me some time to check it. I have to figure out the topology and find a better way.

inochisa commented:
I suggest adding the sg2002 DTS for me to reference the usage.
Also, I have rebased  the for-next to v6.19-rc1, I think you should resolve it.

stavinsky commented:
@inochisa
I've started to implement internal codec driver (adc for now). I came up with the fact that the simple-audio-card is not enough to describe all possible configurations for our AIAO subsystem.  (at least there is no way to request s16_le over 32 bit i2s slot which is the format for internal adc)
It probably means that we have to implement our own "machine driver" to properly connect cpu, codec, and mux. 
In that case tdm driver stays the same, but mux will be replaced with sound card machine driver. 

If you don't like this idea we could stay with current mux, but we will have to hardcode i2s slot width to 32, which is most common in my practice. And we will probably be not able to support PDM mode (I dont know if anyone will want to used it at all )

possible configuration could be like this
```
/* i2s nodes remains the same  */
 / {
	sound {
		compatible = "sophgo,cv1800b-aiao";
		model = "cv1800b audio";

		link@0 {
			link-name = "i2s0-internal-adc";
			cpu-dai = <&i2s0>;
			codec-dai = <&int_adc>;

			dai-format = "i2s";
			bitclock-master = <&i2s0>;
			frame-master = <&i2s0>;
			sophgo,sdi-in = <CV180x_INT_ADC>;
		};

		link@1 {
			link-name = "i2s3-internal-dac";
			cpu-dai = <&i2s3>;
			codec-dai = <&int_dac>;

			dai-format = "i2s";
			bitclock-master = <&i2s3>;
			frame-master = <&i2s3>;
			sophgo,sdo-out = <CV180x_INT_DAC>;
		};

		link@2 {
			link-name = "i2s2-ics43432";
			cpu-dai = <&i2s2>;
			codec-dai = <&ics43432>;

			dai-format = "i2s";
			bitclock-master = "cpu";
			frame-master = "cpu";
			sophgo,sdi-in = <CV180x_EXT_I2S2>;
		};
	};
};

int_adc: codec@300a100 {
	compatible = "sophgo,cv1800b-adc";
	reg = <0x0300a100 0x100>;
	#sound-dai-cells = <0>;
	status = "okay";
};

int_dac: codec@300a000 {
	compatible = "sophgo,cv1800b-dac";
	reg = <0x0300a000 0x100>;
	#sound-dai-cells = <0>;
	status = "okay";
};

codec0: my-adc {
	compatible = "invensense,ics43432";
	#sound-dai-cells = <0>;
};

```

inochisa commented:
> I've started to implement internal codec driver (adc for now). I came up with the fact that the simple-audio-card is not enough to describe all possible configurations for our AIAO subsystem. (at least there is no way to request s16_le over 32 bit i2s slot which is the format for internal adc) It probably means that we have to implement our own "machine driver" to properly connect cpu, codec, and mux. In that case tdm driver stays the same, but mux will be replaced with sound card machine driver.
> 
> If you don't like this idea we could stay with current mux, but we will have to hardcode i2s slot width to 32, which is most common in my practice. 

~~I think it is worthy to switch to the machine driver.~~ I prefer to keep SoC driver. 
But in fact, I do not like the current mux. It is works, but is too ugly. And I doubt the subsystem maintainer will agree as you used a very cold framework. I also want to combine them as one device, but it is also a mess.

> And we will probably be not able to support PDM mode (I dont know if anyone will want to used it at all )

It is fine to leave it if it is hard to implement, but I suggest leaving room for the PDM mode, it is a pain to refactor everything at the time we want to support it. In some times, it means the driver itself should be abandoned and a new one should be added.

PS1: how about using `simple-audio-card,aux-devs` to controll the mux?

inochisa commented on arch/riscv/boot/dts/sophgo/cv180x.dtsi:
```diff
@@ -448,6 +450,60 @@
+
+		i2s0: i2s@4100000 {
+			compatible = "sophgo,cv1800b-i2s";
+			reg = <0x04100000 0x10000>;
```
`reg = <0x04100000 0x8000>;` The mux is at the last 0x8000.

stavinsky replied:
true. Found this problem yesterday.  it should be 1000 not 10000. Typo.

inochisa commented:
Some bad (or not so bad) thoughs:
1. switch mux driver to a syscon device, and let all i2s, pcm, dac.adc drivers access this device directly via regmap. The property can be something like `sophgo,audio-control`
2. Allow to set the I2S mode in the DTS again, use something like `sophgo,i2s-tx-only` and `sophgo,i2s-rx-only`, the default is txrx if no one is set. Since the i2s driver know its mode, it then can setup in a right way instead of enable all of it.

stavinsky commented:
> ~I think it is worthy to switch to the machine driver.~ I prefer to keep SoC driver. But in fact, I do not like the current mux. It is works, but is too ugly. And I doubt the subsystem maintainer will agree as you used a very cold framework. I also want to combine them as one device, but it is also a mess.
> 
> > And we will probably be not able to support PDM mode (I dont know if anyone will want to used it at all )
> 
> It is fine to leave it if it is hard to implement, but I suggest leaving room for the PDM mode, it is a pain to refactor everything at the time we want to support it. In some times, it means the driver itself should be abandoned and a new one should be added.
> 
> PS1: how about using `simple-audio-card,aux-devs` to controll the mux?

I think I can write sound card driver in a few days. This means we will have possibility to configure everything we have in hardware.
So the mux will be replaced with the new sound card driver. (probably we will have even less amount of code for now)
The structure I've described before looks unfinished but it is already describing everything we need in future.  (i2s, pdm, internal codecs)
so the plan is: 
step 1: i2s driver that we already have without mux settings at all in dts (the card driver will be a glue between i2s and codecs as intended)
step 2: sound card driver with ability to connect external/internal codecs via DAI link (easy part if we implement only basics, alsa have all needed callbacks, I just need to implement it in TDM module)
step 3: internal codec (possibly another patch series)
step 4: pdm, pcm formats for full implementation of the driver (another patch series)
step 5: DAPM - but i'm not sure I have enough energy to do that part

If you agree wth the plan, i will make this pull request as draft and come back to you when the card driver will be ready. 
The plan eliminates your not so bad thoughts and we will make it as ALSA recommends. 
But I still need your thoughts about proposed DT for the card driver.

inochisa commented:
> I think I can write sound card driver in a few days. This means we will have possibility to configure everything we have in hardware. 
> So the mux will be replaced with the new sound card driver. (probably we will have even less amount of code for now) 

Great

> The structure I've described before looks unfinished but it is already describing everything we need in future. (i2s, pdm, internal codecs) so the plan is: 
> step 1: i2s driver that we already have without mux settings at all in dts (the card driver will be a glue between i2s and codecs as intended) 

Agreed

> step 2: sound card driver with ability to connect external/internal codecs via DAI link (easy part if we implement only basics, alsa have all needed callbacks, I just need to implement it in TDM module) 

OK, I think it is reasonable, It seems like better to leave the connection in the board.

> step 3: internal codec (possibly another patch series) 

Another is fine. Just be simple

> step 4: pdm, pcm formats for full implementation of the driver (another patch series) 

It is OK for me.

> step 5: DAPM - but i'm not sure I have enough energy to do that part

Anyway, it is enough for something, if you feel too hard for it, I will take the rest. Thanks for your energy :).

> 
> If you agree wth the plan, i will make this pull request as draft and come back to you when the card driver will be ready. 

I think it is good plan, let's start it.

> The plan eliminates your not so bad thoughts and we will make it as ALSA recommends. 

It is good to follow ALSA, and I think it is necessary to keep simple.

> But I still need your thoughts about proposed DT for the card driver.

In fact, it is kind of  relaxed for devicetree, In my view, I prefer a extensible design, but if it is too hard, leave it and just make it worked.

stavinsky commented:
> In fact, it is kind of relaxed for devicetree, In my view, I prefer a extensible design, but if it is too hard, leave it and just make it worked.

I will try to replicate dt as much as possible in the same way as simple-audio-card does with only 1 addition: mux connections. This mostly why we need our card.

stavinsky committed:
dt-bindings: sound: sophgo: add CV1800B I2S/TDM controller binding

Purpose: introduce DT schema for the CPU driver

Signed-off-by: Anton D. Stavinskii <stavinsky@gmail.com>

stavinsky committed:
ASoC: sophgo: add CV1800B I2S/TDM controller driver

The actual CPU DAI controller.

Signed-off-by: Anton D. Stavinskii <stavinsky@gmail.com>

stavinsky committed:
ASoC: sophgo: add CV1800B internal ADC codec driver

Codec DAI endpoint for TXDAC + basic controls.

Signed-off-by: Anton D. Stavinskii <stavinsky@gmail.com>

stavinsky committed:
ASoC: sophgo: add CV1800B internal DAC codec driver

Codec DAI endpoint for TXDAC + basic controls.

Signed-off-by: Anton D. Stavinskii <stavinsky@gmail.com>

stavinsky committed:
riscv: dts: sophgo: dts nodes for i2s tdm modules

New dts nodes for TDMs and readable DMA names

Signed-off-by: Anton D. Stavinskii <stavinsky@gmail.com>

stavinsky committed:
riscv: dts: sophgo: add Milk-V Duo 256M board dts

Example of usage internal dac/adc and external I2S mic

Signed-off-by: Anton D. Stavinskii <stavinsky@gmail.com>

stavinsky committed:
dt-bindings: sound: sophgo: add CV1800B internal ADC codec

Document the internal ADC audio codec integrated in the Sophgo
CV1800B SoC.

Signed-off-by: Anton D. Stavinskii <stavinsky@gmail.com>

stavinsky committed:
dt-bindings: sound: sophgo: add CV1800B internal DAC codec

Document the internal DAC audio codec integrated in the Sophgo
CV1800B SoC.

Signed-off-by: Anton D. Stavinskii <stavinsky@gmail.com>

stavinsky commented:
@inochisa Hey. Hope you are doing well. 

I made a decision to get rid of mux driver in this commit. 
by default it is allocated in the following order:
TDM0 -> internal adc
TDM1 -> external I2S1
TDM2 -> external I2S2
TDM3 -> internal dac 
So I've just described everything with simple-audio-card in the same way. 

Internal ADC/DAC now works out of the box. I2S2 with external mic still need these settings for ethernet IP block. You said it should be done by mdio but mdio is not available yet.

### What we have now 
tdm almost the same. Added a few callbacks to be compatible with simple-card
adc codec - works only on 48kHz 2 channels 
dac codec - works only on 48kHz 2 channels

### Testing and results
can be tested like this. Ideally with  Duo Module 01 EVB board as it have standard jack and both channels available.
on Milk Duo 256M capture also works with left channel, playback was not tested. 
```
 modprobe cv1800b-sound-dac dyndbg=+pt && \
 modprobe cv1800b-sound-adc dyndbg=+pt && \
 modprobe cv1800b-tdm  dyndbg=+pt && \
 modprobe snd-soc-simple-card dyndbg=+pt

amixer sset 'Internal I2S' 12
arecord -D hw:0,1 -f S16_LE -c 2 -r 48000  -d 10 -t wav > dump.wav
aplay  -D hw:0,0 -f S16_LE -c 2 -r 48000  dump.wav
 ```


My external I2S mic was also tested on Milk 256M and it works fine with 32 bit slot on I2S2 (ethernet pins from the side)
For internal DAC/ADC it is important to use provided configuration in DT example specially clocks and rates. It is important because internal ADC uses TDM3 MCLK as master clock. So it should be enabled and with predefined frequency. 

all the commits were checked with 
```
./scripts/checkpatch.pl --subjective --strict -g HEAD
make DT_CHECKER_FLAGS=-m dt_binding_check
make DT_CHECKER_FLAGS=-m W=1 dtbs_check
```
As always appreciate for you support and attention to details. 
Cheers :)


P.S. i've ordered PCM mic to try implement PCM mode. But this is a different story.
Right now only I2S will be available.

inochisa commented on Documentation/devicetree/bindings/sound/sophgo,cv1800b-i2s.yaml:
```diff
@@ -0,0 +1,87 @@
+    description: Optional mux-states consumer property for audio routing.
+
+  mux-state-names:
+    $ref: /schemas/types.yaml#/definitions/string-array
```
This is not needed anymore

inochisa commented on Documentation/devicetree/bindings/sound/sophgo,cv1800b-i2s.yaml:
```diff
@@ -0,0 +1,87 @@
+        compatible = "sophgo,cv1800b-i2s";
+        reg = <0x04110000 0x10000>;
+        #address-cells = <1>;
+        #size-cells = <0>;// check if can be dropped
```
I suggest just leave it, also remove the comment,

inochisa commented on Documentation/devicetree/bindings/sound/sophgo,cv1800b-i2s.yaml:
```diff
@@ -0,0 +1,87 @@
+        pinctrl-0 = <&i2s2_cfg>;
+        pinctrl-names = "default";
+        #sound-dai-cells = <0>;
+        status = "okay";
```
not necessary

inochisa commented:
> I made a decision to get rid of mux driver in this commit. by default it is allocated in the following order: TDM0 -> internal adc TDM1 -> external I2S1 TDM2 -> external I2S2 TDM3 -> internal dac So I've just described everything with simple-audio-card in the same way.
> 

How about adding a property "sophgo,mux" syscon node? so we can configure the mux if needed.

> Internal ADC/DAC now works out of the box. I2S2 with external mic still need these settings for ethernet IP block. You said it should be done by mdio but mdio is not available yet.

The ethernet change I think it maybe too dirty to be upstreamed, so I think it is OK to just leave a comment for it

inochisa commented:
@stavinsky In genernal, I think it is good enough and ready to go, feel free to sent it to the LKML and check if there is more advice, thanks for your PR.

stavinsky commented:
> > I made a decision to get rid of mux driver in this commit. by default it is allocated in the following order: TDM0 -> internal adc TDM1 -> external I2S1 TDM2 -> external I2S2 TDM3 -> internal dac So I've just described everything with simple-audio-card in the same way.
> 
> How about adding a property "sophgo,mux" syscon node? so we can configure the mux if needed.

I'm preparing another patch series with custom sound card. Mux, clocks and so on should be there i think.  

> 
> > Internal ADC/DAC now works out of the box. I2S2 with external mic still need these settings for ethernet IP block. You said it should be done by mdio but mdio is not available yet.
> 
> The ethernet change I think it maybe too dirty to be upstreamed, so I think it is OK to just leave a comment for it

yep. Agree.

stavinsky commented:
> @stavinsky In genernal, I think it is good enough and ready to go, feel free to sent it to the LKML and check if there is more advice, thanks for your PR.

Thanks for your help. It was a great journey.  Ive sent to lkml as you can see  :)

inochisa commented:
Since mainline has already taken all the driver patches, close it.
